refactor(ast): remove `AstKind` for `ModuleDeclaration`

camchenry opened the pull request:
- Part of https://github.com/oxc-project/oxc/issues/11490

Similar to how we handled member expressions, I've added a new `enum` that narrows `AstKind` down to just things that are `ModuleDeclaration`s.

camchenry commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12022?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#12022** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12022?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12022?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#11990** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11990?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=12022).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

camchenry commented:
Mangler benchmark is noise, but the other benchmark changes appear to be real. I'm not 100% sure why but it might be because module declarations are very common, and now we're no longer allocating another unnecessary node for them which has a real perf benefit since we do fewer iterations in semantic and in the linter and elsewhere.

<img width="745" alt="Screenshot 2025-07-01 at 9 58 06 PM" src="https://github.com/user-attachments/assets/8dbd75ba-79a5-43bd-bff3-cd03233c1c7d" />
<img width="747" alt="Screenshot 2025-07-01 at 9 58 17 PM" src="https://github.com/user-attachments/assets/39634871-f09f-4fe1-9bce-fba15cb4c67e" />

camc314 commented on crates/oxc_semantic/src/checker/mod.rs:
```diff
@@ -32,8 +32,13 @@ pub fn check<'a>(kind: AstKind<'a>, ctx: &SemanticBuilder<'a>) {
         AstKind::Directive(dir) => js::check_directive(dir, ctx),
-        AstKind::ModuleDeclaration(decl) => {
-            js::check_module_declaration(decl, ctx);
+        // AstKind::ModuleDeclaration(decl) => {
+        //     js::check_module_declaration(decl, ctx);
+        // }
```
```suggestion
```

camc314 reviewed (approved):
:muscle:

camc314 commented:
> Mangler benchmark is noise, but the other benchmark changes appear to be real. I'm not 100% sure why but it might be because module declarations are very common, and now we're no longer allocating another unnecessary node for them which has a real perf benefit since we do fewer iterations in semantic and in the linter and elsewhere.

yep this feels like the most likley reason, files usually have a decent number of `import/export`decls, so avoidint iterating over them (i guess) is making this faster

Dunqing commented on crates/oxc_linter/src/rules/eslint/no_unused_vars/symbol.rs:
```diff
@@ -186,8 +186,11 @@ impl<'a> Symbol<'_, 'a> {
+                    }
                 }
                 AstKind::ExportNamedDeclaration(_) | AstKind::ExportDefaultDeclaration(_) => {
                     return true;
```
There is a duplicate branch, `ModuleDeclaration` has covered `AstKind::ExportNamedDeclaration(_)` and `AstKind::ExportDefaultDeclaration(_)`. 
In addition, I've looked into the code here, this is checks whether a symbol has been exported, and only `ExportNamedDeclaration` and `ExportDefaultDeclaration` can produce a Symbol, so the `module.is_export()` is unnecessary.

camchenry replied:
I tried removing the `is_export`, but it caused some tests to fail, so I've left it in for now.

Dunqing replied:
Ok, I will take a further look at this.

Dunqing reviewed (changes requested):
Looks good! Just two unnecessary branches can be removed, but this has nothing to do with you because that code has been there for a long time.

Dunqing reviewed (approved):
Nice!

Dunqing commented:
### Merge activity

* **Jul 3, 2:51 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

camchenry committed:
refactor(ast): remove `AstKind` for `ModuleDeclaration` (#12022)

- Part of https://github.com/oxc-project/oxc/issues/11490

Similar to how we handled member expressions, I've added a new `enum` that narrows `AstKind` down to just things that are `ModuleDeclaration`s.
